### PR TITLE
#184651064 user profile settings

### PR DESCRIPTION
**Pull request by Kress2000**
#### What does this PR do?
This is route ``users/profiles/:id`` for the user who has logged in, to edit, add, (Update) their information.
- Here ensures that the user is able to update his info but not others
- Ensures also that user has login first before accessing his info 
#### How should this be manually tested?
Manually, you can use swagger doc or postman to test, 
- here you go there and log in first
- then after login, you will have access to your settings where at /users/settings path, you will be able to add more info about yourself which are these: 
```
  -  place of living
  - date of birth
  - currency
  - gender
  - tel
  - account number
  -  account name
  - postal code
  - street adress
  - country
``` 
 form userDetail table that references users table
According to the schema.
#### Any background context you want to provide?
Of course as the app, this is a PostgreSQL server, and it checks if the user is logged in first and then you can edit

#### What are the relevant pivotal tracker/Trello stories?
#184651064

### CODE REVIEW

File: src/tests/userDetails.test.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      expect(res.body).to.have.property('error');
+    });
+  });
+  let newUser;
+  let newToken;
+  let newUserId;
+  describe('User Details', function () {
+    before(async function () {
+      await sequelize.sync();
+      await sequelize.getQueryInterface().bulkInsert(
```
Comment (Edcreation) [inline src/tests/userDetails.test.js:109]: Sequelize sync is not recommended

**Review by Edcreation — APPROVED** [review]

File: src/tests/userDetails.test.js (hunk 1 above)
Comment (Edcreation) [inline src/tests/userDetails.test.js:109]: Sequelize sync is not recommended

File: src/database/models/userDetails.model.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  placeOfLiving: {
+    type: Sequelize.STRING,
+  },
+  currency: {
+    type: Sequelize.STRING,
+    defaultValue: 'USD',
+    allowNull: false,
+  },
+  lang: {
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply src/database/models/userDetails.model.js]: done!

File: src/controllers/product.controller.js [hunk 3]
```diff
@@ -1,40 +1,33 @@
...
-  } catch (error) {
+  const { cid } = req.params;
+  const collection = await collectionServices.getCollectionById(cid);
+  if (collection) {
+    const deletedCollection = await collectionServices.deleteCollection(cid);
     return res
-      .status(500)
-      .json({ code: 500, message: 'Collection Not Deleted', error });
+      .status(200)
+      .json({ code: 200, message: 'Collection Deleted', deletedCollection });
```
Comment (Edcreation) [inline src/controllers/product.controller.js:30]: don't remove try and catch. what if there is an error? the server would crash you need trycatch to handle errors.

File: src/controllers/product.controller.js [hunk 4]
```diff
@@ -1,40 +1,33 @@
...
+  const oldCollection = await collectionServices.getCollectionByName(
+    collectionObj.name
+  );
+  if (!oldCollection) {
     const collection = await collectionServices.createCollection(collectionObj);
     return res.status(201).json({
       code: 201,
       message: 'Collection Created.',
       collection,
     });
```
Comment (Edcreation) [inline src/controllers/product.controller.js:19]: There is already a middleware checking whether a collection name exists. there was no need to add that.

File: src/database/migrations/20230407081339-create-collections.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        type: DataTypes.UUID,
+        defaultValue: DataTypes.UUIDV4,
+        primaryKey: true,
+        allowNull: false,
+        unique: true,
+      },
+      userId: {
+        type: Sequelize.UUID,
+        allowNull: false,
+        unique: false,
```
Comment (Edcreation) [inline src/database/migrations/20230407081339-create-collections.js:17]: Why did you add another collection migration?

File: src/controllers/product.controller.js [hunk 6]
```diff
@@ -1,40 +1,33 @@
...
-      req.params.cid
-    );
-    if (deletedCollection) {
-      return res.status(200).json({ code: 200, message: 'Collection Deleted' });
-    }
-  } catch (error) {
+  const { cid } = req.params;
+  const collection = await collectionServices.getCollectionById(cid);
+  if (collection) {
+    const deletedCollection = await collectionServices.deleteCollection(cid);
```
Comment (Edcreation) [inline src/controllers/product.controller.js:27]: There is a middleware for checking the validity of a collection no need for this. check **`routes/products.route.js`**

**Review by Edcreation — CHANGES_REQUESTED** [review]

File: src/services/userDetails.services.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  const userInfo = await UserDetailsModel.create(data);
+  return userInfo;
+}
+async function updateUserDetails(userId, data) {
+  const userInfo = await UserDetailsModel.update(data, {
+    where: { userId },
+  });
+  return userInfo;
+}
+export default { getUserDetailsById, createUserDetails, updateUserDetails };
```
Comment (ishimwe5555) [inline src/services/userDetails.services.js:19]: All these three functions are reusable, you can turn them into services, import the model from there only and call those functions in this file. As in we shouldn't import the model everytime we to change a user profile.

File: src/tests/userDetails.test.js [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+  let token;
+  let user;
+  before(async function () {
+    user = await userServices.getUserByEmail(userInfo.email);
+    const res = await chai.request(app).post('/users/login').send(userInfo);
+    token = res.body.token;
+  });
+
+  after(async function () {
+    await UserDetailsModel.destroy({ where: { userId: user.id } });
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:35]: Here you should use the services you would have created instead of calling the model.

File: src/tests/userDetails.test.js [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    await User.destroy({ where: { id: user.id } });
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:36]: check user.services.js there is a deleteUser service, you should use it here instead of importing the model once again. Code reusability.

File: src/tests/userDetails.test.js [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+          currency: 'USD',
+          lang: 'en',
+          dateOfBirth: '1990-01-01',
+          placeOfLiving: 'New York',
+          tel: '+1234567890',
+          postalCode: '44444',
+          accountNumber: '1234',
+          accountName: 'Equity test',
+          country: 'Rwanda test',
+          streetAdress: 'test',
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:77]: I think you'd add them into one object and send the object, to improve code readability.

File: src/tests/userDetails.test.js [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+        .request(app)
+        .post(`/users/profile/${user.id}`)
+        .set({ Authorization: `Bearer ${token}` })
+        .send({
+          gender: 'Male',
+          currency: 'USD',
+          lang: 'en',
+          dateOfBirth: '1990-01-01',
+          placeOfLiving: 'New York',
+          tel: '+1234567890',
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:96]: Improve Code readability by sending one object instead of a whole lot of key-value pairs.

File: src/tests/userDetails.test.js [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+          gender: '',
+          currency: '',
+          lang: '',
+          dateOfBirth: '1990-01-01',
+          placeOfLiving: 'New York',
+          tel: '+1234567890',
+          postalCode: '0000',
+          accountName: 'Equity',
+          country: 'Rwanda',
+          streetAdress: '',
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:117]: Same case here. code readability.

File: src/tests/userDetails.test.js [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+          currency: 'USD',
+          lang: 'test lang',
+          dateOfBirth: '1990-01-01',
+          placeOfLiving: 'New York',
+          tel: '+870899889',
+          postalCode: '0000',
+          accountNumber: '12334',
+          accountName: 'kim',
+          country: 'tzd',
+          streetAdress: 'KN107',
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:204]: Sometimes you are sending the same keys and values, if you had created one object it would reduce the lines of code and improve readability.

File: src/tests/userDetails.test.js [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+        });
+      expect(res).to.have.status(500);
+      expect(res.body).to.be.a('object');
+      expect(res.body).to.have.property('message', 'Error, update failed');
+      expect(res.body).to.have.property('error', 'Validation error');
+    });
+    after(async function () {
+      await UserDetailsModel.destroy({ where: { userId: newUserId } });
+      await User.destroy({ where: { id: newUserId } });
```
Comment (ishimwe5555) [inline src/tests/userDetails.test.js:295]: use the services here

File: src/services/userDetails.services.js (hunk 7 above)
Comment (Kress2000) [reply src/services/userDetails.services.js:19]: they are already in services

File: src/tests/userDetails.test.js (hunk 9 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:36]: okay

File: src/tests/userDetails.test.js (hunk 8 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:35]: okay

File: src/tests/userDetails.test.js (hunk 10 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:77]: this is neither a problem nor an issue

File: src/controllers/product.controller.js (hunk 4 above)
Comment (Kress2000) [reply src/controllers/product.controller.js:19]: pwovide its name

File: src/database/migrations/20230407081339-create-collections.js (hunk 5 above)
Comment (Kress2000) [reply src/database/migrations/20230407081339-create-collections.js:17]: is it a problem? I just even added somelines to modify the userId

File: src/tests/userDetails.test.js (hunk 10 above)
Comment (ishimwe5555) [reply src/tests/userDetails.test.js:77]: Haha I didn't say it was a problem or an issue, it's just my suggestion, to improve code readability.

File: src/controllers/product.controller.js (hunk 6 above)
Comment (Kress2000) [reply src/controllers/product.controller.js:27]: am not checking for validity, am just getting collection by its id

File: src/tests/userDetails.test.js (hunk 14 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:295]: yes sure

File: src/tests/userDetails.test.js (hunk 12 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:117]: prob here, if you have checked well, am not using the same object properties, some where , am using them as null, just for the sake of testing. so, no need for the variable as it is used once

File: src/tests/userDetails.test.js (hunk 13 above)
Comment (Kress2000) [reply src/tests/userDetails.test.js:205]: not the same

File: src/controllers/product.controller.js (hunk 3 above)
Comment (Kress2000) [reply src/controllers/product.controller.js]: okay sure

**Review by ishimwe5555 — CHANGES_REQUESTED** [review]

**Review by AdrineUWERA — APPROVED** [review]

**Review by didasmbalanya — CHANGES_REQUESTED** [review]
migration files should not be edited

**Review by ishimwe5555 — COMMENTED** [review]
You can remove .vscode from github and add it to your .gitignore, otherwise it looks good to me.

File: .vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (Kress2000) [reply .vscode/settings.json:1]: okay, let me do it, thanks

**Review by ishimwe5555 — APPROVED** [review]
LG

**Comment by Kress2000** [comment]
> Code Looks good to me, make sure all the checks are passing.

done now

File: src/services/userDetails.services.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+async function updateUserDetails(userId, data) {
+  const userInfo = await UserDetailsModel.update(data, {
+    where: { userId },
+  });
+  return userInfo;
+}
+async function deleteUserDetails(userId) {
+  const userInfo = await UserDetailsModel.destroy({ where: { userId } });
+  return userInfo;
+}
```
Comment (munezerobagira) [inline src/services/userDetails.services.js:22]: user can't delete his details instead we can set all of them to null

**Review by munezerobagira — CHANGES_REQUESTED** [review]
